src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.GeneratedFileReplacingCompilationTracker.cs

```diff
@@ -0,0 +1,214 @@
+using System.Diagnostics.CodeAnalysis;
+using System.Threading;
+using System.Threading.Tasks;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal partial class SolutionState
+    {
+        /// <summary>
+        /// An implementation of <see cref="ICompilationTracker"/> that takes a compilation from another compilation tracker and updates it
+        /// to return a generated document with a specific content, regardless of what the generator actually produces. In other words, it says
+        /// "take the compilation this other thing produced, and pretend the generator gave this content, even if it wouldn't."
+        /// </summary>
+        private class GeneratedFileReplacingCompilationTracker : ICompilationTracker
+        {
+            private readonly ICompilationTracker _underlyingTracker;
+            private readonly SourceGeneratedDocumentState _replacedGeneratedDocumentState;
+
+            /// <summary>
+            /// The lazily-produced compilation that has the generated document updated. This is initialized by call to
+            /// <see cref="GetCompilationAsync"/>.
+            /// </summary>
+            [DisallowNull]
+            private Compilation? _compilationWithReplacement;
+
+            public GeneratedFileReplacingCompilationTracker(ICompilationTracker underlyingTracker, SourceGeneratedDocumentState replacementDocumentState)
+            {
+                _underlyingTracker = underlyingTracker;
+                _replacedGeneratedDocumentState = replacementDocumentState;
+            }
+
+            public bool HasCompilation => _underlyingTracker.HasCompilation;
+
+            public ProjectState ProjectState => _underlyingTracker.ProjectState;
+
+            public ICompilationTracker Clone()
+            {
+                return new GeneratedFileReplacingCompilationTracker(_underlyingTracker.Clone(), _replacedGeneratedDocumentState);
+            }
+
+            public bool ContainsAssemblyOrModuleOrDynamic(ISymbol symbol, bool primary)
+            {
+                if (_compilationWithReplacement == null)
+                {
+                    // We don't have a compilation yet, so this couldn't have came from us
+                    return false;
+                }
+                else
+                {
+                    return UnrootedSymbolSet.Create(_compilationWithReplacement).ContainsAssemblyOrModuleOrDynamic(symbol, primary);
+                }
+            }
+
+            public bool? ContainsSymbolsWithNameFromDeclarationOnlyCompilation(Func<string, bool> predicate, SymbolFilter filter, CancellationToken cancellationToken)
+            {
+                // This only needs to be implemented if a feature that operates from a source generated file needs to search for declarations
+                // with other names; those APIs are only used by certain code fixes which isn't a need for now.
+                throw new NotImplementedException();
```

Comment:
hrmm... so it's unclear to my why we can't hit this notimpl during normal codepaths.  we have features that search for declarations.   is it because htis only gets created for teh frozen case, and not he normal case.  And since those features aren't freezing docs, they don't hit this?

Reply:
I should have clarified this: at this point we're still not running code fixes on generated files so we can't hit this. We'll fix this up once that's working. I think I have a bug somewhere so I'll update that with a tracking comment.